Added default switch statement

Chairn opened the pull request:
Added default switch statement either with a comment or with a call to dbg_msg. During my tests, i couldn't trigger any of them, but its better to have a default case anyway.

I didn't add default case in https://github.com/ddnet/ddnet/blob/master/src/engine/shared/protocol_ex.cpp#L51, https://github.com/ddnet/ddnet/blob/master/datasrc/compile.py#L266-L281, https://github.com/ddnet/ddnet/blob/master/datasrc/seven/compile.py and https://github.com/ddnet/ddnet/blob/master/src/game/collision.cpp#L183-L215 as the code is either too obscure (protocol) or clear enough (collision).

Tested with gcc 7 and clang 13.

## Checklist

- [x] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test if it works standalone, system.c especially
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

def- commented:
No one reads dbg_msgs, we should probably crash on unexpected values instead of continuing in an undefined state.

Chairn commented:
Yeah, but are you sure they cannot happen? Maybe its possible for a malicious client to forge a packet to trigger an assert where before it would just go on but with an undetermined state.

def- commented:
That's exactly the argument why crashing is better than an undetermined (aka wrong) state. At least that was the argument when we enabled dbg_assert for Release builds too.

Chairn commented:
Some of them are quite obvious for crashing, for instance: https://github.com/ddnet/ddnet/pull/4850/commits/ae23857435e94fd838e408feedfc90270d091b3c#diff-c021a22ad84ce67e915adf5b68a504cb831cdd0abace9f143430422eb678a763R212-R214 but its also inoffensive
Some are inoffensive, like in system.cpp or https://github.com/ddnet/ddnet/pull/4850/commits/ae23857435e94fd838e408feedfc90270d091b3c#diff-50bdad6ef56152ea4b79c252b4484e67b2b7dfe487c3aa0bac7827fd3c9f9d71R221-R223, no need to assert there
but some are trickier, probably because i don't know the codebase well enough, for instance: https://github.com/ddnet/ddnet/pull/4850/commits/ae23857435e94fd838e408feedfc90270d091b3c#diff-619bc2285dfceebaa74c335ad9df122d2cc19e665de4536801ccf63c4ab80714R554 or https://github.com/ddnet/ddnet/pull/4850/commits/ae23857435e94fd838e408feedfc90270d091b3c#diff-0744fff67a75a13690490704148072fde6c44cf1b7a3cedb3d8a0aa35b8ee63bR106-R108

def- commented:
@heinrich5991 any opinion?

Chairn commented:
I can change this to draft, but i'll never be able to fully test the switch in the server part

heinrich5991 commented:
> @heinrich5991 any opinion?

I'm very much in favor of using asserts for states the code shouldn't be in.

heinrich5991 commented:
Status: Waiting for author for changing to `dbg_assert`.

It'd also be nice if a central header like `color.h` didn't include `system.h`.

Chairn commented:
Rebased, moved to dbg_assert, included more switches. Did not include system.h into color.h and added a comment for this specific switch.

Ready for review.

C0D3D3V commented on src/game/editor/explanations.cpp:
```diff
@@ -654,11 +658,13 @@ const char *CEditor::Explain(int ExplanationID, int Tile, int Layer) //TODO: Add
 			return "ENTITIES OFF SIGN: Informs people playing with entities about important marks, tips, information or text on the map.";
 	}
-	return "";
+	return "Unknown or invalid tile";
```
Do we really want this as a default explanation for not set tiles?

Chairn replied:
Better than nothing no?

Chairn replied:
Okay, i reverted this one. Makes more sense to show nothing although trying to place one of these tiles will trigger a warning about invalid tiles.

heinrich5991 commented on src/base/color.h:
```diff
@@ -207,6 +207,8 @@ inline ColorRGBA color_cast(const ColorHSLA &hsl)
 		rgb.b = x;
 		break;
+	default: // leave black color
+		break;
```
This seems like it should assert.

Chairn replied:
I can't use dbg_assert here, is __builtin_trap() alright?

heinrich5991 commented on src/base/system.cpp:
```diff
@@ -3213,9 +3215,10 @@ int str_time(int64_t centisecs, int format, char *buffer, int buffer_size)
 			(centisecs % min) / sec, centisecs % sec);
+	default:
+		dbg_msg("time", "Invalid time formatting data");
+		return -1;
```
assert

Chairn replied:
This one is triggered in test/str.cpp:472

heinrich5991 commented on src/engine/server/server.cpp:
```diff
@@ -1750,6 +1750,9 @@ void CServer::ProcessClientPacket(CNetChunk *pPacket)
 						}
+						default:
+							str_format(aBuf, sizeof(aBuf), "ClientID=%d invalid authentication level=%d", ClientID, AuthLevel);
+							break;
```
assert

Chairn replied:
ok, couldn't trigger this one

heinrich5991 commented on src/game/client/components/emoticon.cpp:
```diff
@@ -208,6 +208,9 @@ void CEmoticon::EyeEmote(int Emote)
 		break;
+	default:
+		str_format(aBuf, sizeof(aBuf), "/emote normal %d", g_Config.m_ClEyeDuration);
+		break;
```
assert

Chairn replied:
I don't think we should assert here as it's harmless for the client.

heinrich5991 replied:
This would be hiding that we got a value that we did not expect. This would be a logic bug and thus we should assert.

heinrich5991 commented on src/game/client/components/menu_background.cpp:
```diff
@@ -218,6 +218,9 @@ void CMenuBackground::LoadMenuBackground(bool HasDayHint, bool HasNightHint)
 				break;
+			default:
+				pMenuMap = "heavens";
+				break;
```
assert

Chairn replied:
I don't think we should assert here as this code path is harmless for the client.

heinrich5991 replied:
Same here, either an assert or a reason why this is the correct and wanted behavior.

heinrich5991 commented on src/game/client/components/players.cpp:
```diff
@@ -626,6 +626,7 @@ void CPlayers::RenderPlayer(
 		case WEAPON_GUN: RenderHand(&RenderInfo, p, Direction, -3 * pi / 4, vec2(-15, 4), Alpha); break;
 		case WEAPON_SHOTGUN: RenderHand(&RenderInfo, p, Direction, -pi / 2, vec2(-5, 4), Alpha); break;
 		case WEAPON_GRENADE: RenderHand(&RenderInfo, p, Direction, -pi / 2, vec2(-4, 7), Alpha); break;
+		default: break; // don't render non projectile weapon
```
What does this have to do with projectiles?

Chairn replied:
Good question, i guess i copied comment from projectile.cpp

heinrich5991 commented on src/game/client/prediction/entities/projectile.cpp:
```diff
@@ -65,6 +65,8 @@ vec2 CProjectile::GetPos(float Time)
 		Speed = pTuning->m_GunSpeed;
 		break;
+	default:
+		break; // ignore non projectile weapons
```
assert

heinrich5991 commented on src/game/client/prediction/gameworld.cpp:
```diff
@@ -614,12 +614,14 @@ CEntity *CGameWorld::FindMatch(int ObjID, int ObjType, const void *pObjData)
 	case NETOBJTYPE_LASER: FindType(ENTTYPE_LASER, CLaser, CNetObj_Laser);
 	case NETOBJTYPE_PICKUP: FindType(ENTTYPE_PICKUP, CPickup, CNetObj_Pickup);
+	default: // no match, return nullptr
+		return nullptr;
```
assert

Chairn replied:
Looks like the default case can happen, i wouldn't assert.

heinrich5991 replied:
I don't see how it can happen. It's only called in few locations, each with controlled `ObjType` parameter.

heinrich5991 commented on src/game/editor/explanations.cpp:
```diff
@@ -654,11 +658,13 @@ const char *CEditor::Explain(int ExplanationID, int Tile, int Layer) //TODO: Add
 			return "ENTITIES OFF SIGN: Informs people playing with entities about important marks, tips, information or text on the map.";
 	}
-	return "";
+	return "Unknown or invalid tile";
```
When is this displayed?

I don't think we should display something for empty tiles.

Chairn replied:
Reverted this one

heinrich5991 commented on src/game/server/entities/projectile.cpp:
```diff
@@ -102,6 +102,8 @@ vec2 CProjectile::GetPos(float Time)
 		}
 		break;
+	default:
+		break; // ignore non projectile weapons
```
assert

heinrich5991 commented on src/game/server/gamecontext.cpp:
```diff
@@ -1267,6 +1267,10 @@ void CGameContext::ProgressVoteOptions(int ClientID)
 		case 13: OptionMsg.m_pDescription13 = pCurrent->m_aDescription; break;
 		case 14: OptionMsg.m_pDescription14 = pCurrent->m_aDescription; break;
+		default:
+			dbg_msg("vote", "Votes to send too big: %d", NumVotesToSend);
```
assert

heinrich5991 commented on src/game/server/player.cpp:
```diff
@@ -910,6 +914,10 @@ void CPlayer::ProcessScoreResult(CScorePlayerResult &Result)
 				GameServer()->SendBroadcast(aBuf, m_ClientID);
 			}
+		}
+			break;
```
Weird formatting.

heinrich5991 reviewed (changes requested):
Still needs more asserts IMO.

Also need a scripts/fix_style.py run.

heinrich5991 commented:
Status: Needs rebase, some review comments remaining.

heinrich5991 commented:
Needs rebase, but @Chairn is currently not interested.
